lib/Event/Traits/RecursionBlockingEventDispatchHelperTrait.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+/**
+ * Pimcore
+ *
+ * This source file is available under two different licenses:
+ * - GNU General Public License version 3 (GPLv3)
+ * - Pimcore Commercial License (PCL)
+ * Full copyright and license information is available in
+ * LICENSE.md which is distributed with this source code.
+ *
+ * @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ * @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Event\Traits;
+
+/**
+ * @internal
+ */
+trait RecursionBlockingEventDispatchHelperTrait
+{
+
+    /**
+     * @var array
+     */
+    private array $activeDispatchingEvents = [];
+
+    /**
+     * Dispatches an event, avoids recursion by checking if the active dispatch event is the same
+     *
+     * @param object $event
+     * @param string|null $eventName
+     * @return void
+     */
+    protected function dispatchEvent(object $event, string $eventName = null): void
+    {
+        $eventName ??= \get_class($event);
+        if (!isset($this->activeDispatchingEvents[$eventName])) {
```

Comment:
What if you have depending events? So if you have a listener that saves another DataObject that should then trigger the same event just for the other object? That might not work with that solution right?

Reply:
That should work, since the dispatching is just blocked for the originating object as it isn't a static variable.